File: algorithm_cpp/sort/mergesort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <ctime>
#include <cstdlib>

using namespace std;

class Mergesort {
private:
    vector<int> aux;            // 辅助数组
public:
    void swap(vector<int> &vec, int i, int j) {
        int temp = vec[i];
        vec[i] = vec[j];
        vec[j] = temp;
    }

    void merge(vector<int> &vec, int lo, int mid, int hi) {
        aux = vec;
        int i = lo, j = mid+1;
        for (int k = lo ; k <= hi ; ++k) {
            if (i > mid) vec[k] = aux[j++];
            else if (j > hi) vec[k] = aux[i++];
            else if (aux[i] <= aux[j])   vec[k] = aux[i++];
            else    vec[k] = aux[j++];
        }
    }

    void sort_helper(vector<int> &vec, int lo, int hi) {
        if (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            sort_helper(vec, lo, mid);
            sort_helper(vec, mid+1, hi);
            merge(vec, lo, mid, hi);
        }
    }

    void sort(vector<int> &vec) {
        int lo = 0, hi = vec.size()-1;
        sort_helper(vec, lo, hi);
    }
};

int main() {
    vector<int> vec(50,0);
    srand(time(NULL));
    generate(vec.begin(), vec.end(), [](){ return rand()%100; });
    for_each(vec.begin(), vec.end(), [](int x){ cout << x << " "; });
    cout << endl;
    Mergesort sol;
    sol.sort(vec);
    for_each(vec.begin(), vec.end(), [](int x){ cout << x << " "; });
    cout << endl;
    return 0;
}

File: algorithm_java/sort/MergeSort.java
package chapter_2_sort;

/**
 * Created by lookfiresu on 16-2-23.
 */
/* 自顶向下归并排序 */
public class MergeSort extends BaseSort {
    private static Comparable[] aux;    // 归并所需的辅助数组

    // 原地归并算法
    public static void merge(Comparable[] a, int lo, int mid, int hi) {
        // 将a[lo:mid]和a[mid+1:hi]归并
        int i = lo, j = mid+1;
        for (int k = lo ; k <= hi ; k++)    // 将a[lo:hi]复制到aux[lo:hi]
            aux[k] = a[k];
        for (int k = lo ; k <= hi ; k++) {
            if (i > mid)    // 若aux[lo:mid]已归并结束，则将aux[mid+1:hi]的剩余部分逐一归并到a中
                a[k] = aux[j++];
            else if (j > hi)    // 若aux[mid+1:hi]已归并结束，则将aux[lo:mid]的剩余部分逐一归并到a中
                a[k] = aux [i++];
            else if (less(aux[j], aux[i]))  // 若aux[j]小于aux[i]，则将aux[j]拷贝到a[k]中，并将a[j+1]加入下次比较
                a[k] = aux[j++];
            else    // 若aux[i]不大于aux[j]，则将aux[i]拷贝到a[k]中，并将a[i+1]加入下次比较
                a[k] = aux[i++];
        }
    }

    public static void sort(Comparable[] a, int lo, int hi) {
        // 将数组a[lo:hi]排序
        if (hi <= lo)   // 若当前数组只有一个元素或为空，则不做任何操作
            return;
        int mid = lo + (hi - lo)/2; // 数组a[lo:mid]和a[mid+1:hi]是相邻且连续的
        sort(a, lo, mid);   // 使用归并排序算法将左半边数组a[lo:mid]排序
        sort(a, mid+1, hi); // 使用归并排序算法将右半部数组a[mid:hi]排序
        merge(a, lo, mid, hi);  // 归并已排序子数组a[lo:mid]和a[mid+1:hi]
    }

    public static void sort(Comparable[] a) {
        aux = new Comparable[a.length];     // 一次性分配空间
        sort(a, 0, a.length-1);
    }
}
